File: backend/btagent_backend/api/v1/workflows.py
"""Workflow CRUD + version lifecycle API (Phase 2 v1).

The route layer is thin — input validation via Pydantic, RBAC + org
scoping, and translation of service-layer ``ValueError`` into 409. All
mutation flows through :mod:`btagent_backend.services.workflow_service`
so the single-published-version-per-workflow invariant lives in one
place.
"""

from __future__ import annotations

import logging

from btagent_shared.types.config import TLP
from btagent_shared.types.workflow import (
    CreateWorkflowRequest,
    CreateWorkflowVersionRequest,
    RunWorkflowRequest,
    UpdateWorkflowRequest,
    UpdateWorkflowVersionRequest,
    WorkflowListResponse,
    WorkflowResponse,
    WorkflowRunListResponse,
    WorkflowRunResponse,
    WorkflowRunStatus,
    WorkflowVersionListResponse,
    WorkflowVersionResponse,
    WorkflowVersionState,
)
from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from btagent_backend.api.deps import CurrentUser, get_current_user, get_db
from btagent_backend.db.models_workflow import WorkflowRow, WorkflowRunRow, WorkflowVersionRow
from btagent_backend.services import workflow_run_service, workflow_service
from btagent_backend.services.workflow_run_service import WorkflowNotExecutable

logger = logging.getLogger("btagent.api.workflows")

router = APIRouter(prefix="/workflows", tags=["workflows"])


# --------------------------------------------------------------------------- #
# Helpers
# --------------------------------------------------------------------------- #


def _to_workflow_response(row: WorkflowRow) -> WorkflowResponse:
    return WorkflowResponse(
        id=row.id,
        name=row.name,
        description=row.description,
        org_id=row.org_id,
        created_by=row.created_by,
        created_at=row.created_at,
        updated_at=row.updated_at,
    )


def _to_version_response(row: WorkflowVersionRow) -> WorkflowVersionResponse:
    return WorkflowVersionResponse(
        id=row.id,
        workflow_id=row.workflow_id,
        version_number=row.version_number,
        state=WorkflowVersionState(row.state),
        definition=row.definition or {},
        org_id=row.org_id,
        created_by=row.created_by,
        created_at=row.created_at,
        published_at=row.published_at,
        deprecated_at=row.deprecated_at,
    )


async def _load_workflow_scoped(
    db: AsyncSession, workflow_id: str, user: CurrentUser
) -> WorkflowRow:
    """Fetch a workflow and 404 if either missing or not in caller's org.

    Returning 404 (not 403) for the cross-tenant case is intentional —
    we don't want to confirm-by-status-code that a workflow id exists in
    another tenant (matches the IDOR-mitigation pattern used by the
    investigations / IOC endpoints).
    """
    wf = await workflow_service.get_workflow(db, workflow_id)
    if wf is None or wf.org_id != user.org_id:
        raise HTTPException(status_code=404, detail="Workflow not found")
    return wf


async def _load_version_scoped(
    db: AsyncSession, workflow: WorkflowRow, version_number: int
) -> WorkflowVersionRow:
    version = await workflow_service.get_version(
        db, workflow_id=workflow.id, version_number=version_number
    )
    if version is None:
        raise HTTPException(status_code=404, detail="Workflow version not found")
    return version


def _to_run_response(row: WorkflowRunRow) -> WorkflowRunResponse:
    return WorkflowRunResponse(
        id=row.id,
        workflow_id=row.workflow_id,
        version_id=row.version_id,
        version_number=row.version_number,
        org_id=row.org_id,
        triggered_by=row.triggered_by,
        status=WorkflowRunStatus(row.status),
        trigger_payload=row.trigger_payload or {},
        outputs=row.outputs or {},
        final_output=row.final_output,
        nodes_executed=row.nodes_executed or [],
        evidence_chain=row.evidence_chain or [],
        error=row.error,
        created_at=row.created_at,
        completed_at=row.completed_at,
    )


# --------------------------------------------------------------------------- #
# Workflow CRUD
# --------------------------------------------------------------------------- #


@router.post("", response_model=WorkflowResponse, status_code=201)
async def create_workflow(
    body: CreateWorkflowRequest,
    db: AsyncSession = Depends(get_db),
    user: CurrentUser = Depends(get_current_user),
):
    """Create a new workflow + its initial draft version 1."""
    user.require_permission("workflow:create")

    wf, _version = await workflow_service.create_workflow(
        db,
        name=body.name,
        description=body.description,
        org_id=user.org_id,
        created_by=user.id,
        initial_definition=body.definition,
    )
    return _to_workflow_response(wf)


@router.get("", response_model=WorkflowListResponse)
async def list_workflows(
    page: int = Query(1, ge=1),
    page_size: int = Query(50, ge=1, le=200),
    db: AsyncSession = Depends(get_db),
    user: CurrentUser = Depends(get_current_user),
):
    """List workflows in the caller's org (newest first)."""
    user.require_permission("workflow:view")
    rows, total = await workflow_service.list_workflows(
        db, org_id=user.org_id, page=page, page_size=page_size
    )
    return WorkflowListResponse(
        items=[_to_workflow_response(r) for r in rows],
        total=total,
    )


@router.get("/{workflow_id}", response_model=WorkflowResponse)
async def get_workflow(
    workflow_id: str,
    db: AsyncSession = Depends(get_db),
    user: CurrentUser = Depends(get_current_user),
):
    user.require_permission("workflow:view")
    wf = await _load_workflow_scoped(db, workflow_id, user)
    return _to_workflow_response(wf)


@router.patch("/{workflow_id}", response_model=WorkflowResponse)
async def update_workflow(
    workflow_id: str,
    body: UpdateWorkflowRequest,
    db: AsyncSession = Depends(get_db),
    user: CurrentUser = Depends(get_current_user),
):
    """Patch workflow metadata (name / description only). Definition lives on versions."""
    user.require_permission("workflow:edit")
    wf = await _load_workflow_scoped(db, workflow_id, user)
    wf = await workflow_service.update_workflow_metadata(
        db, workflow=wf, name=body.name, description=body.description
    )
    return _to_workflow_response(wf)


# --------------------------------------------------------------------------- #
# Version CRUD
# --------------------------------------------------------------------------- #


@router.get("/{workflow_id}/versions", response_model=WorkflowVersionListResponse)
async def list_versions(
    workflow_id: str,
    db: AsyncSession = Depends(get_db),
    user: CurrentUser = Depends(get_current_user),
):
    user.require_permission("workflow:view")
    wf = await _load_workflow_scoped(db, workflow_id, user)
    rows = await workflow_service.list_versions(db, workflow_id=wf.id)
    return WorkflowVersionListResponse(
        items=[_to_version_response(r) for r in rows],
        total=len(rows),
    )


@router.get(
    "/{workflow_id}/versions/{version_number}",
    response_model=WorkflowVersionResponse,
)
async def get_version(
    workflow_id: str,
    version_number: int,
    db: AsyncSession = Depends(get_db),
    user: CurrentUser = Depends(get_current_user),
):
    user.require_permission("workflow:view")
    wf = await _load_workflow_scoped(db, workflow_id, user)
    version = await _load_version_scoped(db, wf, version_number)
    return _to_version_response(version)


@router.post(
    "/{workflow_id}/versions",
    response_model=WorkflowVersionResponse,
    status_code=201,
)
async def create_version(
    workflow_id: str,
    body: CreateWorkflowVersionRequest,
    db: AsyncSession = Depends(get_db),
    user: CurrentUser = Depends(get_current_user),
):
    """Stage a new draft version of the workflow.

    Auto-assigned ``version_number = max(existing) + 1``.
    """
    user.require_permission("workflow:edit")
    wf = await _load_workflow_scoped(db, workflow_id, user)
    version = await workflow_service.create_version(
        db,
        workflow=wf,
        definition=body.definition,
        created_by=user.id,
    )
    return _to_version_response(version)


@router.patch(
    "/{workflow_id}/versions/{version_number}",
    response_model=WorkflowVersionResponse,
)
async def update_version(
    workflow_id: str,
    version_number: int,
    body: UpdateWorkflowVersionRequest,
    db: AsyncSession = Depends(get_db),
    user: CurrentUser = Depends(get_current_user),
):
    """Patch a draft version's definition. 409 if already published."""
    user.require_permission("workflow:edit")
    wf = await _load_workflow_scoped(db, workflow_id, user)
    version = await _load_version_scoped(db, wf, version_number)
    try:
        version = await workflow_service.update_version_definition(
            db, version=version, definition=body.definition
        )
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc))
    return _to_version_response(version)


# --------------------------------------------------------------------------- #
# Lifecycle transitions
# --------------------------------------------------------------------------- #


@router.post(
    "/{workflow_id}/versions/{version_number}/publish",
    response_model=WorkflowVersionResponse,
)
async def publish_version(
    workflow_id: str,
    version_number: int,
    db: AsyncSession = Depends(get_db),
    user: CurrentUser = Depends(get_current_user),
):
    """Promote a draft → published. Auto-deprecates the prior published version."""
    user.require_permission("workflow:publish")
    wf = await _load_workflow_scoped(db, workflow_id, user)
    version = await _load_version_scoped(db, wf, version_number)
    try:
        version = await workflow_service.publish_version(db, version=version)
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc))
    return _to_version_response(version)


@router.post(
    "/{workflow_id}/versions/{version_number}/deprecate",
    response_model=WorkflowVersionResponse,
)
async def deprecate_version(
    workflow_id: str,
    version_number: int,
    db: AsyncSession = Depends(get_db),
    user: CurrentUser = Depends(get_current_user),
):
    """Explicit deprecate (admin path). 409 if the version is still a draft."""
    user.require_permission("workflow:deprecate")
    wf = await _load_workflow_scoped(db, workflow_id, user)
    version = await _load_version_scoped(db, wf, version_number)
    try:
        version = await workflow_service.deprecate_version(db, version=version)
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc))
    return _to_version_response(version)


# --------------------------------------------------------------------------- #
# Execution + run history
# --------------------------------------------------------------------------- #


@router.post(
    "/{workflow_id}/versions/{version_number}/run",
    response_model=WorkflowRunResponse,
    status_code=201,
)
async def run_version(
    workflow_id: str,
    version_number: int,
    body: RunWorkflowRequest,
    db: AsyncSession = Depends(get_db),
    user: CurrentUser = Depends(get_current_user),
):
    """Execute a workflow version and persist a run record (UC: run API).

    Runs synchronously and returns the terminal run (``succeeded`` /
    ``failed`` / ``paused``). A 201 is returned even for a ``failed`` run:
    the execution attempt produced a durable record. 422 is reserved for
    the case where the version isn't a runnable graph at all (empty /
    malformed definition).

    **TLP fail-closed:** when ``body.active_tlp`` is omitted we default to
    ``TLP.RED`` (most restrictive) so an unconfigured caller cannot bypass
    the per-capability egress gate by silence. Callers triggering a run
    from a classified investigation must pass that investigation's
    classification on the request body.
    """
    user.require_permission("workflow:run")
    wf = await _load_workflow_scoped(db, workflow_id, user)
    version = await _load_version_scoped(db, wf, version_number)
    try:
        run = await workflow_run_service.execute_version(
            db,
            workflow=wf,
            version=version,
            trigger_payload=body.trigger_payload,
            triggered_by=user.id,
            active_tlp=body.active_tlp if body.active_tlp is not None else TLP.RED,
        )
    except WorkflowNotExecutable as exc:
        raise HTTPException(status_code=422, detail=str(exc))
    return _to_run_response(run)


@router.get("/{workflow_id}/runs", response_model=WorkflowRunListResponse)
async def list_runs(
    workflow_id: str,
    page: int = Query(1, ge=1),
    page_size: int = Query(50, ge=1, le=200),
    db: AsyncSession = Depends(get_db),
    user: CurrentUser = Depends(get_current_user),
):
    """List execution history for a workflow (newest first)."""
    user.require_permission("workflow:view")
    wf = await _load_workflow_scoped(db, workflow_id, user)
    rows, total = await workflow_run_service.list_runs(
        db, workflow_id=wf.id, page=page, page_size=page_size
    )
    return WorkflowRunListResponse(items=[_to_run_response(r) for r in rows], total=total)


@router.get("/{workflow_id}/runs/{run_id}", response_model=WorkflowRunResponse)
async def get_run(
    workflow_id: str,
    run_id: str,
    db: AsyncSession = Depends(get_db),
    user: CurrentUser = Depends(get_current_user),
):
    """Fetch a single run record."""
    user.require_permission("workflow:view")
    wf = await _load_workflow_scoped(db, workflow_id, user)
    run = await workflow_run_service.get_run(db, run_id=run_id)
    # 404 (not 403) on cross-workflow / cross-tenant access — same posture
    # as the workflow IDOR mitigation above.
    if run is None or run.workflow_id != wf.id or run.org_id != user.org_id:
        raise HTTPException(status_code=404, detail="Workflow run not found")
    return _to_run_response(run)

File: frontend/src/api/workflows.ts
import api from "./client";

// --------------------------------------------------------------------------- //
// Types — mirror shared/btagent_shared/types/workflow.py + the engine
// Workflow graph shape (compiler/workflow.py) stored in version.definition.
// --------------------------------------------------------------------------- //

export type WorkflowVersionState = "draft" | "published" | "deprecated";

export type WorkflowRunStatus =
  | "pending"
  | "running"
  | "succeeded"
  | "failed"
  | "paused";

export type TLP = "red" | "amber_strict" | "amber" | "green" | "white";

/** A node in the compiled engine Workflow graph (version.definition.nodes[]). */
export interface WorkflowGraphNode {
  step_id: string;
  node_id: string;
  name: string;
  config: Record<string, unknown>;
}

/** An edge in the compiled engine Workflow graph (version.definition.edges[]). */
export interface WorkflowGraphEdge {
  source: string;
  target: string;
  label: string;
}

/** The engine Workflow serialized into version.definition (may be {} for an empty draft). */
export interface WorkflowDefinition {
  name?: string;
  version?: string;
  description?: string;
  trigger?: Record<string, unknown>;
  nodes?: WorkflowGraphNode[];
  edges?: WorkflowGraphEdge[];
}

export interface Workflow {
  id: string;
  name: string;
  description: string;
  org_id: string;
  created_by: string | null;
  created_at: string;
  updated_at: string;
}

export interface WorkflowVersion {
  id: string;
  workflow_id: string;
  version_number: number;
  state: WorkflowVersionState;
  definition: WorkflowDefinition;
  org_id: string;
  created_by: string | null;
  created_at: string;
  published_at: string | null;
  deprecated_at: string | null;
}

/** One hash-linked audit entry (engine EvidenceRecord JSON). */
export interface EvidenceRecord {
  run_id: string;
  node_id: string;
  prev_hash: string;
  link_hash: string;
  input_hash: string;
  output_hash: string;
  timestamp: string;
}

export interface WorkflowRun {
  id: string;
  workflow_id: string;
  version_id: string;
  version_number: number;
  org_id: string;
  triggered_by: string | null;
  investigation_id: string | null;
  status: WorkflowRunStatus;
  /** When paused, the step id awaiting approval. */
  paused_node_id: string | null;
  /** Step ids approved across resume cycles. */
  approved_steps: string[];
  trigger_payload: Record<string, unknown>;
  outputs: Record<string, unknown>;
  final_output: Record<string, unknown> | null;
  nodes_executed: string[];
  evidence_chain: EvidenceRecord[];
  error: string | null;
  created_at: string;
  completed_at: string | null;
}

export interface WorkflowListResponse {
  items: Workflow[];
  total: number;
}

export interface WorkflowVersionListResponse {
  items: WorkflowVersion[];
  total: number;
}

export interface WorkflowRunListResponse {
  items: WorkflowRun[];
  total: number;
}

export interface CreateWorkflowRequest {
  name: string;
  description?: string;
  definition?: WorkflowDefinition;
}

export interface UpdateWorkflowRequest {
  name?: string;
  description?: string;
}

export interface CreateWorkflowVersionRequest {
  definition: WorkflowDefinition;
}

export interface UpdateWorkflowVersionRequest {
  definition: WorkflowDefinition;
}

export interface RunWorkflowRequest {
  trigger_payload?: Record<string, unknown>;
  /** Omit to fail-closed at TLP.RED; or inherit from investigation_id. */
  active_tlp?: TLP;
  investigation_id?: string;
}

// --------------------------------------------------------------------------- //
// Workflow CRUD
// --------------------------------------------------------------------------- //

export async function listWorkflows(
  params: { page?: number; page_size?: number } = {},
): Promise<WorkflowListResponse> {
  const sp = new URLSearchParams();
  if (params.page) sp.set("page", String(params.page));
  if (params.page_size) sp.set("page_size", String(params.page_size));
  const q = sp.toString();
  return api.get<WorkflowListResponse>(`/v1/workflows${q ? `?${q}` : ""}`);
}

export async function getWorkflow(id: string): Promise<Workflow> {
  return api.get<Workflow>(`/v1/workflows/${id}`);
}

export async function createWorkflow(data: CreateWorkflowRequest): Promise<Workflow> {
  return api.post<Workflow>("/v1/workflows", data);
}

export async function updateWorkflow(
  id: string,
  data: UpdateWorkflowRequest,
): Promise<Workflow> {
  return api.patch<Workflow>(`/v1/workflows/${id}`, data);
}

// --------------------------------------------------------------------------- //
// Versions
// --------------------------------------------------------------------------- //

export async function listVersions(workflowId: string): Promise<WorkflowVersionListResponse> {
  return api.get<WorkflowVersionListResponse>(`/v1/workflows/${workflowId}/versions`);
}

export async function getVersion(
  workflowId: string,
  versionNumber: number,
): Promise<WorkflowVersion> {
  return api.get<WorkflowVersion>(`/v1/workflows/${workflowId}/versions/${versionNumber}`);
}

export async function createVersion(
  workflowId: string,
  data: CreateWorkflowVersionRequest,
): Promise<WorkflowVersion> {
  return api.post<WorkflowVersion>(`/v1/workflows/${workflowId}/versions`, data);
}

export async function updateVersion(
  workflowId: string,
  versionNumber: number,
  data: UpdateWorkflowVersionRequest,
): Promise<WorkflowVersion> {
  return api.patch<WorkflowVersion>(
    `/v1/workflows/${workflowId}/versions/${versionNumber}`,
    data,
  );
}

export async function publishVersion(
  workflowId: string,
  versionNumber: number,
): Promise<WorkflowVersion> {
  return api.post<WorkflowVersion>(
    `/v1/workflows/${workflowId}/versions/${versionNumber}/publish`,
  );
}

export async function deprecateVersion(
  workflowId: string,
  versionNumber: number,
): Promise<WorkflowVersion> {
  return api.post<WorkflowVersion>(
    `/v1/workflows/${workflowId}/versions/${versionNumber}/deprecate`,
  );
}

// --------------------------------------------------------------------------- //
// Node catalog (canvas palette)
// --------------------------------------------------------------------------- //

export type NodeCategory =
  | "trigger"
  | "integration"
  | "reasoning"
  | "knowledge"
  | "decision"
  | "data"
  | "output";

export interface NodeCatalogEntry {
  id: string;
  name: string;
  version: string;
  category: NodeCategory | string;
  description: string;
  /** JSON Schema of the node's pydantic input model; {} when none. */
  input_schema: Record<string, unknown>;
}

export interface NodeCatalogResponse {
  items: NodeCatalogEntry[];
  total: number;
}

export async function getNodeCatalog(): Promise<NodeCatalogResponse> {
  return api.get<NodeCatalogResponse>("/v1/workflows/node-catalog");
}

// --------------------------------------------------------------------------- //
// Execution + run history
// --------------------------------------------------------------------------- //

export async function runVersion(
  workflowId: string,
  versionNumber: number,
  data: RunWorkflowRequest = {},
): Promise<WorkflowRun> {
  return api.post<WorkflowRun>(
    `/v1/workflows/${workflowId}/versions/${versionNumber}/run`,
    data,
  );
}

export async function listRuns(
  workflowId: string,
  params: { page?: number; page_size?: number } = {},
): Promise<WorkflowRunListResponse> {
  const sp = new URLSearchParams();
  if (params.page) sp.set("page", String(params.page));
  if (params.page_size) sp.set("page_size", String(params.page_size));
  const q = sp.toString();
  return api.get<WorkflowRunListResponse>(
    `/v1/workflows/${workflowId}/runs${q ? `?${q}` : ""}`,
  );
}

export async function getRun(workflowId: string, runId: string): Promise<WorkflowRun> {
  return api.get<WorkflowRun>(`/v1/workflows/${workflowId}/runs/${runId}`);
}

/** Approve a paused run's gate and resume it (requires hitl:approve). */
export async function resumeRun(workflowId: string, runId: string): Promise<WorkflowRun> {
  return api.post<WorkflowRun>(`/v1/workflows/${workflowId}/runs/${runId}/resume`);
}
